tests/suite/test_transport_server_tcp_load_balance.py

```diff
@@ -0,0 +1,114 @@
+    delete_service,
+    scale_deployment
+)
+from suite.custom_resources_utils import (
+    patch_ts
+)
+
+
+@pytest.mark.ts
+@pytest.mark.parametrize(
+    "crd_ingress_controller, transport_server_setup",
+    [
+        (
+            {
+                "type": "complete",
+                "extra_args":
+                    [
+                        "-global-configuration=nginx-ingress/nginx-configuration",
+                        "-enable-leader-election=false"
+                    ]
+            },
+            {"example": "transport-server-tcp-load-balance", "app_type": "simple"},
+        )
+    ],
+    indirect=True,
+)
+class TestTransportServerTcpLoadBalance:
+
+    def restore_transport_server(self, kube_apis, transport_server_setup) -> None:
+        """
+        Function to revert a TransportServer resource to a valid state.
+        """
+        transport_server_file = f"{TEST_DATA}/transport-server-status/standard/transport-server.yaml"
+        patch_ts(kube_apis.custom_objects, transport_server_setup.name, transport_server_file, transport_server_setup.namespace)
+
+    @pytest.mark.sean
+    def test_number_of_replicas(
+        self, kube_apis, crd_ingress_controller, transport_server_setup, ingress_controller_prerequisites
+    ):
+        """
+        The load balancing of TCP should result in 4 servers to match the 4 replicas of a service.
+        """
+        original = scale_deployment(kube_apis.apps_v1_api, "tcp-service", transport_server_setup.namespace, 4)
+
+        wait_before_test()
+
+        result_conf = get_ts_nginx_template_conf(
+            kube_apis.v1,
+            transport_server_setup.namespace,
+            transport_server_setup.name,
+            transport_server_setup.ingress_pod_name,
+            ingress_controller_prerequisites.namespace
+        )
+
+        pattern = 'server .*;'
+        num_servers = len(re.findall(pattern, result_conf))
+
+        assert num_servers is 4
+
+        scale_deployment(kube_apis.apps_v1_api, "tcp-service", transport_server_setup.namespace, original)
+
+    def test_tcp_request_load_balanced(
+            self, kube_apis, crd_ingress_controller, transport_server_setup
+    ):
+        """
+        Requests to the load balanced TCP service should result in responses from 3 different endpoints.
+        """
+        node_port_file = f"{TEST_DATA}/transport-server-tcp-load-balance/standard/node-port.yaml"
+        node_port_name = create_service_from_yaml(
+            kube_apis.v1,
+            transport_server_setup.namespace,
+            node_port_file,
+        )
+
+        wait_before_test(3)
+
+        node_port_service = kube_apis.v1.read_namespaced_service(node_port_name, transport_server_setup.namespace)
+        port = node_port_service.spec.ports[0].node_port
+        host = transport_server_setup.public_endpoint.public_ip
+
+        print(f"sending tcp requests to: {host}:{port}")
+
+        endpoints = {}
+        for i in range(20):
```

Comment:
in this test, we're not really making connections through NGINX in this test 🤔  perhaps I'm missing something?

the NodePort port needs to be configured in the Ingress Controller service. After configuring it, we can send a connection to the `public_endpoint.public_ip` and the nodeport. This way the connection will go through NGINX.

Currently, the connection goes like this:
```
client -> node -> kube-proxy -> backend
```
meaning we're not testing NGINX

Reply:
That's frustrating as this was the intent of the test!

I haven't been able to get it to work. Do you know how it should be setup?

Reply:
It's not clear what you mean when you say: `the NodePort port needs to be configured in the Ingress Controller service.`
Do you mean the Nginx-ingress deployment `deployments/deployment/nginx-ingress.yaml` ?
I think that's used by the tests `tests.suite.resources_utils.create_ingress_controller`

Or do you mean the `kind: service` with the name `nginx-ingress` here: `deployments/service/nodeport.yaml` ?

Reply:
Okay I think I have it:
```
get services -A
NAMESPACE                      NAME            TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                                                                   AGE
nginx-ingress                  nginx-ingress   NodePort    10.106.78.48    <none>        80:31426/TCP,443:31001/TCP,8080:32493/TCP,9113:31972/TCP,3333:32249/TCP   12s
test-namespace-1619176893926   tcp-service     ClusterIP   10.107.12.201   <none>        3333/TCP
```

```
k get endpoints -A
NAMESPACE                      NAME                       ENDPOINTS                                                     AGE
nginx-ingress                  nginx-ingress              172.17.0.3:9113,172.17.0.3:3333,172.17.0.3:8080 + 2 more...   73s
test-namespace-1619176893926   tcp-service                172.17.0.4:3333,172.17.0.5:3333,172.17.0.6:3333               68s
```

```
k describe service nginx-ingress -n nginx-ingress
Name:                     nginx-ingress
Namespace:                nginx-ingress
...
Port:                     tcp-server  3333/TCP
TargetPort:               3333/TCP
NodePort:                 tcp-server  32249/TCP
Endpoints:                172.17.0.3:3333
```
I think this means endpoint is pointing to the ingress-controller endpoint `172.17.0.3`.
If we contact the public endpoint, we get the tcp-service IPs
```
nc 192.168.64.2 32249
172.17.0.4:3333
nc 192.168.64.2 32249
172.17.0.5:3333%
```

So I think it's going:
```
public endpoint (192.168.64.2 32249) -> ingress controller (172.17.0.3:3333) -> tcp-server (172.17.0.4:3333)
```